In [531]:
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
import seaborn as sns
import statsmodels.api as sm
import logging

from sklearn.preprocessing import StandardScaler

from scipy.stats import skew, kurtosis

from utils.general import *

import json

import warnings

In [532]:
warnings.filterwarnings('ignore')

In [533]:
random_state = 42
rng = np.random.default_rng(random_state)

# Data Loading And Preprocessing

In [534]:
billboard_df = pd.read_csv("../data/billboard/hot-100_all.csv")
spotify_df = pd.read_csv("../data/spofity/songs.csv")
audio_analysis_df = pd.read_csv("../data/audio/audio_features_full.csv")

In [535]:
billboard_df.drop(axis=1, inplace=True, labels=['image', 'artist'])
billboard_df['date'] = pd.to_datetime(billboard_df['date'])
spotify_df.drop(axis=1, inplace=True, labels=['spotify_name',
                                              'artist',
                                              'artist_genres',
                                              'spotify_id',
                                              'spotify_id',
                                              'spotify_uri',
                                              'spotify_external_url',
                                              'spotify_artist_popularity',
                                              'preview_url',
                                              'preview_url_audio',
                                              'full_audio',
                                              'full_audio_duration_s'
                                              ])
audio_analysis_df.drop(axis=1, inplace=True, labels=['name', 'tempo'])

In [536]:
songs_df = spotify_df.merge(audio_analysis_df, how='inner', on='billboard_name')
songs_df

billboard_name  duration_ms  spotify_popularity  \
0         All I Want For Christmas Is You     241106.0                69.0   
1       Rockin' Around The Christmas Tree     126266.0                62.0   
2                        Jingle Bell Rock     130973.0                62.0   
3                 A Holly Jolly Christmas     135533.0                54.0   
4                                 Circles     215280.0                86.0   
...                                   ...          ...                 ...   
1418  Christmas Isn't Canceled (Just You)     231549.0                41.0   
1419                       Moved To Miami     222225.0                66.0   
1420                              Hibachi     170413.0                69.0   
1421                             Thailand     200958.0                70.0   
1422                          Do It To It     157890.0                91.0   

      spotify_artist_popularity_mean explicit  danceability  energy   key  \
0                          81.000000    False         0.336   0.627   7.0   
1                          59.000000    False         0.589   0.472   8.0   
2                          50.000000    False         0.754   0.424   2.0   
3                          48.000000    False         0.683   0.375   0.0   
4                          91.000000    False         0.695   0.762   0.0   
...                              ...      ...           ...     ...   ...   
1418                       77.000000    False         0.580   0.789   1.0   
1419                       88.500000     True         0.717   0.444   1.0   
1420                       86.333333     True         0.681   0.522   5.0   
1421                       84.000000     True         0.875   0.478   7.0   
1422                       78.000000    False         0.854   0.806  11.0   

      loudness  mode  ...  mfcc16_mean  mfcc16_var  mfcc17_mean  mfcc17_var  \
0       -7.463   1.0  ...     1.619618   99.735660    -3.865835   99.024666   
1       -8.749   1.0  ...    -1.039626   78.420586    -4.437555   55.536427   
2       -8.463   1.0  ...     1.430321   58.685158    -4.030815   67.332291   
3      -13.056   1.0  ...    -1.355817   60.197350    -6.695084   52.782772   
4       -3.497   1.0  ...     2.563944   78.141319   -12.359889   83.661438   
...        ...   ...  ...          ...         ...          ...         ...   
1418    -4.918   0.0  ...     5.321839   70.412506     1.213545   77.698616   
1419   -11.126   1.0  ...     4.598643  180.801086    -4.373017  110.878738   
1420    -8.740   0.0  ...     8.091636   79.854568    -1.646704  123.382797   
1421   -10.562   1.0  ...    10.420262   91.743813    -2.071233   75.112267   
1422    -8.262   0.0  ...     2.155907   68.733521    -5.703623   47.707085   

      mfcc18_mean  mfcc18_var  mfcc19_mean  mfcc19_var  mfcc20_mean  \
0        2.729783  112.219580    -7.488522  122.569650     2.323859   
1        3.890496   70.359543     0.014326   77.899239     6.889563   
2        1.802275   58.469532    -5.335912   53.423290     0.133941   
3       -4.325858   66.221947    -3.533713   50.849602    -1.266797   
4        4.207565   65.643173    -5.280680   54.441185    -0.751733   
...           ...         ...          ...         ...          ...   
1418     4.693950   85.284431    -2.604682   76.687698     5.805956   
1419     4.545245  111.550697     4.918246   63.780304     8.114554   
1420     5.309008   98.507568     0.213412   80.767159     2.804790   
1421     8.341298   97.730263    -0.260812   69.875168     1.959964   
1422    -1.433100   45.689129    -6.835608   45.040623    -1.940495   

      mfcc20_var  
0     141.572560  
1      93.610161  
2      58.774597  
3      90.991325  
4      59.799530  
...          ...  
1418   77.433144  
1419   75.806396  
1420   73.490234  
1421   62.722679  
1422   51.794838  

[1423 rows x 75 columns]

In [537]:
# print(songs_df.apply(lambda x: x.nunique()))
songs_df.describe()

duration_ms  spotify_popularity  spotify_artist_popularity_mean  \
count    1423.000000         1423.000000                     1423.000000   
mean   200696.557976           67.404076                       83.302977   
std     50770.975125           17.661345                       12.018877   
min     32000.000000            0.000000                        0.000000   
25%    170322.000000           64.000000                       78.000000   
50%    195428.000000           70.000000                       86.000000   
75%    223599.000000           77.000000                       91.000000   
max    613026.000000           95.000000                      100.000000   

       danceability       energy          key     loudness         mode  \
count   1423.000000  1423.000000  1423.000000  1423.000000  1423.000000   
mean       0.666045     0.622016     5.153197    -6.747289     0.622628   
std        0.151579     0.162841     3.600340     2.579960     0.484900   
min        0.150000     0.007600     0.000000   -33.663000     0.000000   
25%        0.570000     0.525000     1.000000    -7.841500     0.000000   
50%        0.680000     0.633000     5.000000    -6.360000     1.000000   
75%        0.776000     0.733500     8.000000    -5.077500     1.000000   
max        0.965000     0.984000    11.000000    -1.321000     1.000000   

       speechiness  acousticness  ...  mfcc16_mean   mfcc16_var  mfcc17_mean  \
count  1423.000000   1423.000000  ...  1423.000000  1423.000000  1423.000000   
mean      0.139837      0.223816  ...     2.398114    84.513799    -3.647905   
std       0.125819      0.253148  ...     3.631675    24.471115     3.398574   
min       0.023200      0.000003  ...   -13.240079    32.670311   -18.392536   
25%       0.042800      0.031550  ...     0.246056    68.009289    -5.805181   
50%       0.079800      0.121000  ...     2.442511    81.295982    -3.595455   
75%       0.218000      0.321500  ...     4.722763    98.500587    -1.416037   
max       0.699000      0.995000  ...    13.629806   207.025589     6.795540   

        mfcc17_var  mfcc18_mean   mfcc18_var  mfcc19_mean   mfcc19_var  \
count  1423.000000  1423.000000  1423.000000  1423.000000  1423.000000   
mean     82.115186     2.668189    80.818220    -2.539943    79.075432   
std      24.637780     3.253561    24.665541     3.023733    24.767819   
min      31.161884   -10.105947    31.325922   -14.229393    28.560005   
25%      65.208897     0.716595    64.160015    -4.468847    62.009989   
50%      78.402687     2.753472    77.774918    -2.510375    75.437561   
75%      94.838615     4.810408    93.118820    -0.625247    91.067223   
max     219.371109    12.521308   229.869766     7.686097   229.040588   

       mfcc20_mean   mfcc20_var  
count  1423.000000  1423.000000  
mean      1.677588    80.371770  
std       3.186344    26.740148  
min     -13.992401    28.597084  
25%      -0.239844    61.936447  
50%       1.633557    75.101379  
75%       3.652482    93.510078  
max      14.772246   226.710175  

[8 rows x 72 columns]

In [538]:
songs_df.isna().sum()

billboard_name                    0
duration_ms                       0
spotify_popularity                0
spotify_artist_popularity_mean    0
explicit                          0
                                 ..
mfcc18_var                        0
mfcc19_mean                       0
mfcc19_var                        0
mfcc20_mean                       0
mfcc20_var                        0
Length: 75, dtype: int64

In [539]:
songs_df = pd.get_dummies(songs_df, prefix=['explicit'], columns=['explicit'])

# Feature Engineering

### I don't know what made me think we need *MORE* features, but here's more features...
Some extra feature engineering for spotify's audio analysis object that did't happen during dataset construction.

In [540]:
# this code is so slow it hurts me...

def spotify_audio_analysis(filename):
    file = open(filename)
    analysis_json = json.load(file)
    file.close()
    timbre_coefs = np.arange(13, 1, -1)
    bars = analysis_json['bars']
    beats = analysis_json['beats']
    sections = analysis_json['sections']
    segments = analysis_json['segments']
    tatums = analysis_json['tatums']

    return pd.DataFrame(data={
        # "time_signature": [analysis_json['track']['time_signature']],
        # "mode": [analysis_json['track']['mode']],
        "num_bars": len(bars),
        "mean_bar_duration": np.mean([b['duration'] for b in bars]),
        "var_bar_duration": np.var([b['duration'] for b in bars]),
        "num_beats": len(beats),
        "mean_beat_duration": np.mean([b['duration'] for b in beats]),
        "var_beat_duration": np.var([b['duration'] for b in beats]),
        "num_sections": len(sections),
        "mean_section_duration": np.mean([s['duration'] for s in sections]),
        "var_section_duration": np.var([s['duration'] for s in sections]),
        "mean_section_tempo": np.mean([s['tempo'] for s in sections]),
        "var_section_tempo": np.var([s['tempo'] for s in sections]),
        "mean_section_loudness": np.mean([s['loudness'] for s in sections]),
        "var_section_loudness": np.var([s['loudness'] for s in sections]),
        # other section stuff
        "num_segments": len(segments),
        "mean_segment_duration": np.mean([s['duration'] for s in segments]),
        "var_segment_duration": np.var([s['duration'] for s in segments]),
        "mean_segment_loudness_start": np.mean([s['loudness_start'] for s in segments]),
        "var_segment_loudness_start": np.var([s['loudness_start'] for s in segments]),
        "mean_segment_loudness_max": np.mean([s['loudness_max'] for s in segments]),
        "var_segment_loudness_max": np.var([s['loudness_max'] for s in segments]),
        "mean_segment_loudness_max_time": np.mean([s['loudness_max_time'] for s in segments]),
        "var_segment_loudness_max_time": np.var([s['loudness_max_time'] for s in segments]),
        "mean_segment_loudness_end": np.mean([s['loudness_end'] for s in segments]),
        "var_segment_loudness_end": np.var([s['loudness_end'] for s in segments]),
        "mean_segment_loudness_start_max_diff": np.mean([abs(s['loudness_start'] - s['loudness_max']) for s in segments]),
        "var_segment_loudness_start_max_diff": np.var([abs(s['loudness_start'] - s['loudness_max']) for s in segments]),
        "mean_segment_loudness_start_end_diff": np.mean([abs(s['loudness_start'] - s['loudness_end']) for s in segments]),
        "var_segment_loudness_start_end_diff": np.var([abs(s['loudness_start'] - s['loudness_end']) for s in segments]),
        "mean_segment_loudness_max_end_diff": np.mean([abs(s['loudness_max'] - s['loudness_end']) for s in segments]),
        "var_segment_loudness_max_end_diff": np.var([abs(s['loudness_max'] - s['loudness_end']) for s in segments]),
        "mean_segment_num_pitches": np.mean([len(s['pitches']) for s in segments]),
        "var_segment_num_pitches": np.var([len(s['pitches']) for s in segments]),
        "mean_segment_num_pure_pitches": np.mean([len(np.array(s['pitches'])[np.array(s['pitches']) > 0.5]) for s in segments]),
        "var_segment_num_pure_pitches": np.var([len(np.array(s['pitches'])[np.array(s['pitches']) > 0.5]) for s in segments]),
        "mean_segment_timbre": np.mean([np.dot(s['timbre'], timbre_coefs) for s in segments]),
        "var_segment_timbre": np.var([np.dot(s['timbre'], timbre_coefs) for s in segments]),
        "num_tatums": len(tatums),
        "mean_tatum_duration": np.mean([t['duration'] for t in tatums]),
        "var_tatum_duration": np.var([t['duration'] for t in tatums]),
    }, index=[0])

##### This will take a little bit to run

In [541]:
aa = pd.DataFrame()
for i in range(len(songs_df)):
    aa = pd.concat([aa, spotify_audio_analysis("../data/spofity/audio_analysis/" + spotify_df['audio_analysis_file'].iloc[0])], axis=0)
aa.reset_index(inplace=True)
songs_df = pd.concat([songs_df, aa], axis=1)
songs_df.drop(labels='index', axis=1, inplace=True)

In [542]:
songs_df.describe()

duration_ms  spotify_popularity  spotify_artist_popularity_mean  \
count    1423.000000         1423.000000                     1423.000000   
mean   200696.557976           67.404076                       83.302977   
std     50770.975125           17.661345                       12.018877   
min     32000.000000            0.000000                        0.000000   
25%    170322.000000           64.000000                       78.000000   
50%    195428.000000           70.000000                       86.000000   
75%    223599.000000           77.000000                       91.000000   
max    613026.000000           95.000000                      100.000000   

       danceability       energy          key     loudness         mode  \
count   1423.000000  1423.000000  1423.000000  1423.000000  1423.000000   
mean       0.666045     0.622016     5.153197    -6.747289     0.622628   
std        0.151579     0.162841     3.600340     2.579960     0.484900   
min        0.150000     0.007600     0.000000   -33.663000     0.000000   
25%        0.570000     0.525000     1.000000    -7.841500     0.000000   
50%        0.680000     0.633000     5.000000    -6.360000     1.000000   
75%        0.776000     0.733500     8.000000    -5.077500     1.000000   
max        0.965000     0.984000    11.000000    -1.321000     1.000000   

       speechiness  acousticness  ...  var_segment_loudness_max_end_diff  \
count  1423.000000   1423.000000  ...                       1.423000e+03   
mean      0.139837      0.223816  ...                       4.533173e+01   
std       0.125819      0.253148  ...                       8.316273e-13   
min       0.023200      0.000003  ...                       4.533173e+01   
25%       0.042800      0.031550  ...                       4.533173e+01   
50%       0.079800      0.121000  ...                       4.533173e+01   
75%       0.218000      0.321500  ...                       4.533173e+01   
max       0.699000      0.995000  ...                       4.533173e+01   

       mean_segment_num_pitches  var_segment_num_pitches  \
count                    1423.0                   1423.0   
mean                       12.0                      0.0   
std                         0.0                      0.0   
min                        12.0                      0.0   
25%                        12.0                      0.0   
50%                        12.0                      0.0   
75%                        12.0                      0.0   
max                        12.0                      0.0   

       mean_segment_num_pure_pitches  var_segment_num_pure_pitches  \
count                   1.423000e+03                  1.423000e+03   
mean                    2.263092e+00                  2.966942e+00   
std                     3.331840e-14                  6.663680e-14   
min                     2.263092e+00                  2.966942e+00   
25%                     2.263092e+00                  2.966942e+00   
50%                     2.263092e+00                  2.966942e+00   
75%                     2.263092e+00                  2.966942e+00   
max                     2.263092e+00                  2.966942e+00   

       mean_segment_timbre  var_segment_timbre  num_tatums  \
count         1.423000e+03        1.423000e+03      1423.0   
mean          1.864666e+03        1.180701e+06      1106.0   
std           2.956897e-12        2.049630e-08         0.0   
min           1.864666e+03        1.180701e+06      1106.0   
25%           1.864666e+03        1.180701e+06      1106.0   
50%           1.864666e+03        1.180701e+06      1106.0   
75%           1.864666e+03        1.180701e+06      1106.0   
max           1.864666e+03        1.180701e+06      1106.0   

       mean_tatum_duration  var_tatum_duration  
count         1.423000e+03        1.423000e+03  
mean          2.157633e-01        6.160453e-03  
std           2.832064e-15        7.201633e-17  
min           2.157633e-01        6.

### Testing Different Popularity Metrics

In [568]:
def popularity_metrics(df, score_type='basic'):
    metrics = ['peak_rank', 'debut_rank', 'lifetime_peak_rank', 'sensationality', 'avg_rank_score', 'std_rank_score', 'time_on_chart', 'num_occurrences', 'rank_sum', 'skewness', 'kurtosis', 'spotify_popularity']
    metric_vals = []

    for i in range(len(df)):
        song = df.iloc[i]
        billboard_entries = billboard_df[billboard_df['title'] == song['billboard_name']].sort_values(by='date')
        value_counts = billboard_entries['rank'].value_counts()
        ranks = value_counts.index
        rank_counts = value_counts.values

        # max rank
        peak = min(ranks)
        if score_type == 'classic':
            ranks = [rank_score_classic(peak, ra) for ra in ranks]
            scaled = False
        elif score_type == 'score_01':
            ranks = [rank_score_01(peak, ra) for ra in ranks]
            scaled = True
        elif score_type == 'score_02':
            ranks = [rank_score_02(peak, ra) for ra in ranks]
            scaled = True
        else:
            ranks = [rank_score_basic(ra) for ra in ranks]
            scaled = True

        # sensationality
        sensation = squiggle(rank_counts, ranks, scaled=scaled)

        # mean rank
        avg_rank = np.mean(ranks)

        # rank std
        std_rank = np.std(ranks)

        # length
        time_on_chart = np.max(billboard_entries['weeks'].values)

        # lifetime_peak
        lifetime_peak = np.min(billboard_entries['peakPos'].values)

        # debut rank
        debut_rank = billboard_entries['rank'].iloc[0]

        # number of occurrences
        num_occurrences = len(billboard_entries)

        rank_sum = sum(ranks)

        skewness = skew(ranks)

        kurt = kurtosis(ranks)

        metric_vals.append([peak, debut_rank, lifetime_peak, sensation, avg_rank, std_rank, time_on_chart, num_occurrences, rank_sum, skewness, kurt, song['spotify_popularity']])

    return pd.DataFrame(data=metric_vals, columns=metrics)

In [569]:
pop_metrics_basic = popularity_metrics(songs_df, 'basic')
pop_metrics_basic

peak_rank  debut_rank  lifetime_peak_rank  sensationality  \
0             1           1                   1        7.161467   
1             2           2                   2        3.779657   
2             3           3                   3        2.326703   
3             4           4                   4        1.668480   
4             1           5                   1        7.369801   
...         ...         ...                 ...             ...   
1418         79          79                  79        0.012658   
1419         85          85                  85        0.011765   
1420         91          91                  91        0.010989   
1421         95          95                  95        0.010526   
1422         97          97                  97        0.010309   

      avg_rank_score  std_rank_score  time_on_chart  num_occurrences  \
0           0.270183        0.316913             50               14   
1           0.123908        0.156764             44               17   
2           0.151254        0.099182             41               12   
3           0.121060        0.075093             25               11   
4           0.153094        0.215097             61               47   
...              ...             ...            ...              ...   
1418        0.012658        0.000000              1                1   
1419        0.011765        0.000000              1                1   
1420        0.010989        0.000000              1                1   
1421        0.010526        0.000000              1                1   
1422        0.010309        0.000000              1                1   

      rank_sum  skewness  kurtosis  spotify_popularity  
0     2.161467  1.395650  0.673814                69.0  
1     1.362990  1.352697  0.450922                62.0  
2     1.210036  0.477332 -0.985764                62.0  
3     0.968480  0.378995 -1.307941                54.0  
4     2.908793  3.182407  9.711951                86.0  
...        ...       ...       ...                 ...  
1418  0.012658  0.000000 -3.000000                41.0  
1419  0.011765  0.000000 -3.000000                66.0  
1420  0.010989  0.000000 -3.000000                69.0  
1421  0.010526  0.000000 -3.000000                70.0  
1422  0.010309  0.000000 -3.000000                91.0  

[1423 rows x 12 columns]

In [545]:
# pop_metrics_classic = popularity_metrics(songs_df, 'classic')
# pop_metrics_classic.describe()

In [546]:
# pop_metrics_score_01 = popularity_metrics(songs_df, 'score_01')
# pop_metrics_score_01.describe()

In [547]:
# pop_metrics_score_02 = popularity_metrics(songs_df, 'score_02')
# pop_metrics_score_02.describe()

In [ ]:
data_df = songs_df.copy()
song_billboard_names = data_df.pop('billboard_name')
song_audio_analysis_files = data_df.pop('audio_analysis_file')
data_df.drop(labels=['spotify_popularity', 'spotify_artist_popularity_mean'], axis=1, inplace=True)
data_df[data_df.columns] = StandardScaler().fit_transform(data_df)
# data_df.to_csv("../data/data.csv")
# pop_metrics_basic.to_csv("../data/popularity_metrics.csv")

In [ ]:
pop_metrics = pop_metrics_basic
pop_metrics[pop_metrics.columns] = StandardScaler().fit_transform(pop_metrics)

sns.distplot(pop_metrics['skewness'])
pp_x = data_df
pp_met = pop_metrics
pop_metrics

In [ ]:
from sklearn.utils import shuffle

pop_metrics['sen_range'] = pop_metrics['avg_rank_score'].apply(lambda x: 1 if x >= 0.5 else 0)
pp_met_pos = pop_metrics[pop_metrics['sen_range'] == 1]
pp_met_neg = shuffle(pop_metrics[pop_metrics['sen_range'] == 0]).head(len(pp_met_pos))
pp_met = pd.concat([pp_met_pos, pp_met_neg])
pp_x = data_df.iloc[pp_met.index]

sns.distplot(pp_met['avg_rank_score'])

# sns.countplot(pp_met['sen_range'])
# y_col = 'sensationality'
# y = data_df.pop(y_col)

In [ ]:
x = np.sort(pop_metrics['skewness'])
f_x = np.gradient(x)*np.exp(-x**2/2)
sample_probs = f_x/np.sum(f_x)
# samples = np.random.choice(x, p=sample_probs, size=1000000)
df_samples = pop_metrics["skewness"].sort_values().sample(
    n=400,
    weights=sample_probs,
    replace=False,
)

sns.distplot(df_samples)
# df_samples
pp_met = pop_metrics.iloc[df_samples.index]
pp_x = data_df.iloc[df_samples.index]
# song_billboard_names = songs_df[df_samples.index]

# pop_metrics[pop_metrics["sensationality"] > 0]

In [ ]:
from sklearn.model_selection import GridSearchCV, KFold, train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.svm import SVC
from sklearn.metrics import mean_squared_error
import itertools
import time


# pop_metrics = [pop_metrics_basic, pop_metrics_classic, pop_metrics_score_01, pop_metrics_score_02]
# for pms in pop_metrics:
#     ln = LinearRegression()
#     param_map = {"y": [pms[pm] for pm in pms.columns]}
#     print(gscv(ln, param_map, data_df.values, score_method='neg_root_mean_squared_error'))

def process_subset(model, X, y, features):
    kf = KFold()
    mse = []
    X = X[list(features)]
    for train_idx, test_idx in kf.split(X, y):
        X_train, X_test = X.iloc[train_idx], X.iloc[test_idx]
        y_train, y_test = y.iloc[train_idx], y.iloc[test_idx]
        regr = model.fit(X_train, y=y_train)
        y_predict = regr.predict(X_test)
        mse.append(mean_squared_error(y_test, y_predict))
    return {"model": model, "features": features, "MSE":np.mean(mse)}

def best_subset(model, X, y, subset_size=-1, force=False):
    limit = 100000
    if subset_size == -1:
        subset_size = len(X.columns)
    assert len(X.columns) >= subset_size, f"X.columns must be >= subset_size. Given len(X.columns)={len(X.columns)}, subset_size={subset_size}"

    num_combinations = np.math.factorial(len(X.columns))/(np.math.factorial(len(X.columns)-subset_size) * np.math.factorial(subset_size))
    if num_combinations > limit and not force:
        print(f"Please be aware that this action run {int(num_combinations)} models and there is a real possibility it will crash your machine.\nIf you're sure you want to do this please include the parameter 'force=True' when calling this function")
        return {"model": model, "features": X.columns, "MSE": 2}
    tic = time.time()
    results = []
    for combo in itertools.combinations(X.columns, subset_size):
        results.append(process_subset(model, X, y, combo))
    models = pd.DataFrame(results)
    best_model = models.loc[models['MSE'].argmin()]
    toc = time.time()
    print("Processed", models.shape[0], "models on", subset_size, "predictors in", (toc-tic), "seconds.")
    return best_model

# def forward(model, X, y, staring_predictors):
#     remaining_predictors = [p for p in X.columns if p not in staring_predictors]
#     tic = time.time()
#     results = []
#     for p in remaining_predictors:
#         results.append(process_subset(model, X, y, staring_predictors+[p]))
#     models = pd.DataFrame(results)
#     best_model = models.loc[models['MSE'].argmin()]
#     toc = time.time()
#     print("Processed ", models.shape[0], "models on", len(predictors)+1, "predictors in", (toc-tic), "seconds.")
#
#     Return the best model, along with some other useful information about the model
    # return best_model

# all_features = ['duration_ms', 'key', 'energy', 'loudness', 'danceability', 'speechiness', 'instrumentalness', 'tempo']
pp_x = pp_x.filter(regex=('^((?!mfcc).)*$'))

best_model = best_subset(LinearRegression(), pp_x, pp_met['avg_rank_score'], -1)
print(f"Best Features: f{best_model['features']}")
x_train, x_test, y_train, y_test = train_test_split(pp_x[list(best_model['features'])], pp_met['avg_rank_score'])
mdl = LinearRegression().fit(x_train, y_train)
y_pred = mdl.predict(x_test)
print(f"mse: {mean_squared_error(y_test, y_pred)}")

fig = plt.figure(figsize=(10,5))
plt.scatter(y_test,y_pred)
plt.plot(y_test,y_test,'r', label='test')
plt.plot(y_pred,y_pred,'b', label='pred')

In [ ]:
data_df.iloc[0]

In [ ]:
from sklearn.decomposition import PCA

pca = PCA(n_components=0.99, svd_solver='full')
pca_data = pca.fit_transform(data_df)

pca_comp = pd.DataFrame(pca.components_, index=pca_data.index)
pca_comp.style.background_gradient(vmin=-1, vmax=1, cmap=sns.color_palette("vlag", as_cmap=True))
pca_comp.index

In [ ]:
pca_comp

In [ ]:
pca_data

1. All 111 or 71 if mfccs are left out
2. 1 model with forward regression
3. 1 model with backward regression
4. best 2 models with best subset with len=avg(forward and backward)
5. Limit our initial feature set => data visualisations and exploration